Sources/NIOCore/NIOScheduledCallback.swift

```diff
@@ -0,0 +1,166 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A type that handles callbacks scheduled with `EventLoop.scheduleCallback(at:handler:)`.
+///
+/// - Seealso: `EventLoop.scheduleCallback(at:handler:)`.
+public protocol NIOScheduledCallbackHandler {
+    /// This function is called at the scheduled time, unless the scheduled callback is cancelled.
+    ///
+    /// - Parameter eventLoop: The event loop on which the callback was scheduled.
+    func handleScheduledCallback(eventLoop: some EventLoop)
+
+    /// This function is called if the scheduled callback is cancelled.
+    ///
+    /// The callback could be cancelled explictily, by the user calling ``NIOScheduledCallback/cancel()``, or
+    /// implicitly, if it was still pending when the event loop was shut down.
+    ///
+    /// - Parameter eventLoop: The event loop on which the callback was scheduled.
+    func onCancelScheduledCallback(eventLoop: some EventLoop)
```

Comment:
Ultra naming nit: Should this be `didCancelScheduledCallback`? This follows how Swift APIs are often called when something did happen. `on` is more used when it is about to happen or in the process of happening. e.g. `viewDidLoad` in UIKit vs `onAppear` in SwiftUI

Reply:
I think that `on` is also used for when something has happened but is more ambiguous about whether it's after. IIRC it's `will` that implies before. E.g. `willSet` and `didSet`.

A quick `git-grep` in the NIO repo shows a use of both `on` and `did` quite a bit.

However, it's fair that the only `public` API uses `did`:

```
Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducerStrategies.swift:37:        public mutating func didYield(bufferDepth: Int) -> Bool {
Sources/NIOCore/AsyncSequences/NIOAsyncSequenceProducerStrategies.swift:42:        public mutating func didConsume(bufferDepth: Int) -> Bool {
Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift:70:    public func didYield(_ element: Element) {
Sources/NIOPosix/PendingDatagramWritesManager.swift:249:    public mutating func didWrite(
Sources/NIOPosix/PendingWritesManager.swift:201:    public mutating func didWrite(
```

I can change to using `did` 😅